src/metabase/db/liquibase.clj

```diff
@@ -197,31 +203,39 @@
   "Check and make sure the database isn't locked. If it is, sleep for 2 seconds and then retry several times. There's a
   chance the lock will end up clearing up so we can run migrations normally."
   [^Liquibase liquibase]
-  (u/auto-retry 5
-    (when (migration-lock-exists? liquibase)
-      (Thread/sleep 2000)
-      (throw
-       (LockException.
-        (str
-         (trs "Database has migration lock; cannot run migrations.")
-         " "
-         (trs "You can force-release these locks by running `java -jar metabase.jar migrate release-locks`.")))))))
+  (if (migration-lock-exists? liquibase)
+    (do
+     (log/info "Database has migration lock. Waiting for lock to be cleared...")
+     (Thread/sleep 2000)
+     (u/auto-retry 4
+       (when (migration-lock-exists? liquibase)
+         (Thread/sleep 2000)
+         (throw
+          (LockException.
+           (str
+            (trs "Database has migration lock; cannot run migrations.")
+            " "
+            (trs "You can force-release these locks by running `java -jar metabase.jar migrate release-locks`.")))))))
+    (log/info "No migration lock found.")))
 
 (defn migrate-up-if-needed!
   "Run any unrun `liquibase` migrations, if needed."
-  [^Liquibase liquibase]
+  [^Liquibase liquibase ^DataSource data-source]
   (log/info (trs "Checking if Database has unrun migrations..."))
-  (if (seq (unrun-migrations liquibase))
+  (if (seq (unrun-migrations data-source))
     (do
-     (log/info (trs "Database has unrun migrations. Waiting for migration lock to be cleared..."))
+     (log/info (trs "Database has unrun migrations. Checking if migraton lock is cleared..."))
      (wait-for-migration-lock-to-be-cleared liquibase)
```

Comment:
A note on lock safety - which should probably block on taking the lock immediately after we see it has cleared. This might be a factor in the race conditions we're seeing.

Reply:
I don't follow what you mean, can you expand on that?

Reply:
Typically in concurrent code you would use an atomic operation to block on taking the lock yourself, rather than just waiting for it to be released. With the current behavior there's nothing to stop another process or server taking the lock after this thread resumes.

Reply:
I'm just pointing this out for now as it's independent of your change, but I'll probably send you a PR to fix it. Even if it's not a problem in practice, it's good to follow best practices with concurrent code rather than relying on global reasoning that make not hold in future.

Reply:
ah! I look forward to that PR of your, thank you for pointing that out!